pkg/infradb/infradb.go

```diff
@@ -0,0 +1,1443 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (c) 2022-2023 Intel Corporation, or its subsidiaries.
+// Copyright (C) 2023 Nordix Foundation.
+
+// Package infradb exposes the interface for the manipulation of the api objects
+package infradb
+
+import (
+	"errors"
+	"log"
+	"sync"
+
+	"github.com/opiproject/opi-evpn-bridge/pkg/infradb/common"
+	"github.com/opiproject/opi-evpn-bridge/pkg/infradb/subscriberframework/eventbus"
+	"github.com/opiproject/opi-evpn-bridge/pkg/infradb/taskmanager"
+	"github.com/opiproject/opi-evpn-bridge/pkg/storage"
+	"github.com/philippgille/gokv"
+)
+
+var infradb *InfraDB
+var globalLock sync.Mutex
+
+// InfraDB structure
+type InfraDB struct {
+	client gokv.Store
+}
+
+var (
+	// ErrKeyNotFound error for missing key
+	ErrKeyNotFound = errors.New("key not found")
+	// ErrComponentNotFound error for missing component
+	ErrComponentNotFound = errors.New("component not found")
+	// ErrVrfNotFound error for missing vrf
+	ErrVrfNotFound = errors.New("the referenced VRF has not been found")
+	// ErrLogicalBridgeNotFound error for missing logical bridge
+	ErrLogicalBridgeNotFound = errors.New("the referenced Logical Bridge has not been found")
+	// ErrVrfNotEmpty vrf is not empty
+	ErrVrfNotEmpty = errors.New("the VRF is not empty")
+	// ErrLogicalBridgeNotEmpty logical bridge is not empty
+	ErrLogicalBridgeNotEmpty = errors.New("the LogicalBridge is not empty")
+	// ErrRoutingTableInUse routing table is in use
+	ErrRoutingTableInUse = errors.New("the routing table is already in use")
+	// ErrVniInUse vni is in use
+	ErrVniInUse = errors.New("the VNI is already in use")
+	// Add more error constants as needed
+)
+
+// NewInfraDB initializes an InfraDB object
+func NewInfraDB(address string, dbtype string) error {
+	store, err := storage.NewStore(dbtype, address)
+	if err != nil {
+		log.Println(err)
+		return err
+	}
+
+	infradb = &InfraDB{
+		client: store.GetClient(),
+	}
+	return nil
+}
+
+// Close closes a infradb connection to the DB
+func Close() error {
+	return infradb.client.Close()
+}
+
+// CreateLB creates an infradb logical bridge object
+func CreateLB(lb *LogicalBridge) error {
+	globalLock.Lock()
+	defer globalLock.Unlock()
+
+	vpns := make(map[uint32]bool)
+
+	subscribers := eventbus.EBus.GetSubscribers("logical-bridge")
+	if subscribers == nil {
+		log.Println("CreateLB(): No subscribers for Logical Bridge objects")
```

Comment:
error since no subscribers?

Reply:
Yes Maybe we should exit the programm if we do not have any subscribers. I will discuss this internally

Reply:
Yes @atulpatel261194 will take care of this

Reply:
Done